WIP-add aws-creds-secret

sallyom opened the pull request:
WIP - will write more l8r

sallyom commented on pkg/asset/manifests/content/tectonic/aws-creds-secret-and-reader-role.go:
```diff
@@ -0,0 +1,31 @@
+  resourceNames: ["aws-creds-secret"]
+  verbs: ["get"]
+`))
+)
```
Not sure if any labels/annotations are required here.

sallyom commented on docs/user/environment-variables.md:
```diff
@@ -38,6 +38,8 @@ The installer accepts a number of environment variable that allow the interactiv
 
+* `AWS_PROFILE`:
+     The AWS profile that corresponds to value in `${HOME}/.aws/credentials`.  If not provided, the default is "default".
 * `OPENSHIFT_INSTALL_AWS_REGION`:
```
If this is not set, aws-sdk-go and/or terraform? defaults to "default".  I currently use "openshift-dev" profile.  exporting `AWS_PROFILE=openshift-dev` results in picking up correct aws credentials.

sallyom commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -55,6 +69,7 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
 	}
 
 	assetData := map[string][]byte{
+		"99_aws-creds-secret-and-reader-role.yaml": applyTemplateData(content.AwsCredsSecretAndReaderRole, templateData),
```
Only want to create this if cloud-provider is "aws" so.. have to figure that out.

wking commented:
> TODO: create secret/role only if cloud provider is aws, and/or make secret more generically 'cloud-creds-secret'.

The libvirt analog is probably the libvird URI (and maybe a client cert/key, #296).

steveeJ commented on docs/user/environment-variables.md:
```diff
@@ -38,6 +38,8 @@ The installer accepts a number of environment variable that allow the interactiv
 
 ## Platform-Specific
 
+* `AWS_PROFILE`:
```
If the `OPENSHIFT_INSTALL_` prefix was used here, as with the other variables, it'd be consistent and also allow for a different profile used with the openshift-installer than what is generally set for the environment. The paid for that is to override `AWS_PROFILE` accordingly before calling out to terraform in the installer and also pass it to the aws-sdk-go in the code somehow.

steveeJ commented on docs/user/environment-variables.md:
```diff
@@ -38,6 +38,8 @@ The installer accepts a number of environment variable that allow the interactiv
 
 ## Platform-Specific
 
+* `AWS_PROFILE`:
```
If the `OPENSHIFT_INSTALL_` prefix was used here, as with the other variables, it'd be consistent and also allow for a different profile used with the openshift-installer than what is generally set for the environment. The paid for that is to override `AWS_PROFILE` accordingly before calling out to terraform in the installer and also pass it to the aws-sdk-go in the code somehow.

wking replied:
> If the `OPENSHIFT_INSTALL_` prefix was used here...

I prefer `AWS_PROFILE` as it stands because it's a widely-used variable.  One less thing for users to learn about before calling the installer.

> ... also allow for a different profile used with the openshift-installer than what is generally set for the environment.

Do we expect this to be frequent?  As it stands, folks can override for just the installer:

```console
$ AWS_PROFILE=whatever openshift-installer
```

Although #336 makes all of this moot at the moment ;).

sallyom replied:
:) looking!

sallyom replied:
from aws-sdk-go docs: 
`By default, the SDK checks the AWS_PROFILE environment variable to determine which profile to use. If no AWS_PROFILE variable is set, the SDK uses the default profile.`

sallyom replied:
more info, I unset AWS_PROFILE to get the exact error: when i run with this PR:
```DEBUG       Generating Tectonic Manifests...       
FATAL Failed to generate Terraform Variables: failed to fetch dependency for Terraform Variables: failed to fetch dependency for Bootstrap Ignition Config: failed to fetch dependency for Common Manifests: failed to generate asset Tectonic Manifests: SharedCredsLoad: failed to get profile
caused by: section 'default' does not exist 
```

steveeJ replied:
> Although #336 makes all of this moot at the moment ;).

Ah that's unfortunate, I did expect this to be passable as an argument to the aws-go-sdk functions. After some digging there's [a way to provide credentials](https://docs.aws.amazon.com/sdk-for-go/api/aws/#Config.WithCredentials) to the SDK.
Although after thinking about this:

> Do we expect this to be frequent? As it stands, folks can override for just the installer: ...

I think the effort is probably not worth it and we can document the usage of `AWS_PROFILE` somewhere.

steveeJ commented:
> The libvirt analog is probably the libvird URI

Does the libvirt URI need to be part of the secret? It's not sensitive information and it's already handled by general cluster information.

wking commented:
> > The libvirt analog is probably the libvird URI
>
> Does the libvirt URI need to be part of the secret?

No, but once you need a client cert/key, the URI won't be any use without the secret key.  It seems easier to just keep the cert/key and URI in the same place.  But none of this needs to happen in this PR; just something to think about when planning ahead.

sallyom commented:
@steveeJ @wking is a multi-cloud cluster possible atm? I'm going with installconfig platform (aws|libvirt|openstack) as the only options.  Is there/will there be an option for more than 1 provider?

wking commented:
>  Is there/will there be an option for more than 1 provider?

Not to my knowledge.  And the installer code currently assumes one provider in a few places, so I'd just roll with that approach for now.

smarterclayton commented:
/retest

staebler commented on pkg/asset/manifests/template.go:
```diff
@@ -1,5 +1,31 @@
+}
+
+type cloudCredsTemplateData struct {
+	AwsCredsData
```
Give explicit names to these fields (e.g., AWS, OpenStack, Libvirt). That will (1) make the name in the template file shorter and (2) allow for fields in the various cred data struct to have the same names.

sallyom replied:
done

flaper87 commented:
@sallyom thanks for including `OpenStack` in this PR. cc'ing @tomassedovic @hardys @russellb too. 

Let us know if we can do anything to help on the `OpenStack` side of things.

sallyom commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -80,6 +121,19 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
 	}
+	switch {
+	case installConfig.Config.Platform.AWS != nil:
+		t.files = remove(t.files, "99_openstack-creds-secret-and-reader-role.yaml")
```
yup, that is silly.. fixing now..

sallyom commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -39,8 +41,44 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
+			return err
+		}
+		awsCreds = AwsCredsData{
+			Base64encodeAWSaccessKeyID:     base64.StdEncoding.EncodeToString([]byte(awscreds.AccessKeyID)),
```
got it, changed.

sallyom commented on pkg/asset/manifests/content/tectonic/aws-creds-secret-and-reader-role.go:
```diff
@@ -0,0 +1,31 @@
+  namespace: kube-system
+  name: cloud-creds-secret
+data:
+  aws_access_key_id: {{.CloudCreds.AwsCredsData.Base64encodeAWSaccessKeyID}}
```
looking into that, makes sense

sallyom commented:
this is from aws cluster installed w/ this PR:
```
$ oc describe secret cloud-creds-secret -n kube-system
Name:         cloud-creds-secret
Namespace:    kube-system
Labels:       <none>
Annotations:  <none>
Type:  Opaque
Data
====
aws_secret_access_key:  40 bytes
aws_access_key_id:      20 bytes
```

staebler commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -39,8 +41,45 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
+		if err != nil {
+			return err
+		}
+		awsCreds = AWS{
```
Instead of using temporary `AWS`, `Libvirt`, and `OpenStack` variables, create `cloudCreds` here.

```
cloudCreds := cloudCredsTemplateData{
    AWS: AWS{
        Base64encodeAccessKeyID:     base64.StdEncoding.EncodeToString([]byte(awscreds.AccessKeyID)),
        Base64encodeSecretAccessKey: base64.StdEncoding.EncodeToString([]byte(awscreds.SecretAccessKey)),
    },
}
```

sallyom replied:
:+1:

sallyom replied:
I still have to initialize that cloudCreds var in order to use it below, unless i'm missing something.  I got rid of the temp AWS/Libvirt/OpenStack vars

staebler commented on pkg/asset/manifests/template.go:
```diff
@@ -1,5 +1,31 @@
 package manifests
 
+// AWS holds encoded credentials and is used to generate cloud-creds-secret
+type AWS struct {
```
The previous names of these types were better because they offered a better description of what they were. As it is, a type of `AWS` in the `pkg/asset/manifests` package is not very clear.

sallyom replied:
yup, i misunderstood you yesterday, got it

staebler commented on pkg/asset/manifests/template.go:
```diff
@@ -1,5 +1,31 @@
+}
+
+type cloudCredsTemplateData struct {
+	AWS
```
Explicitly name these fields. You are not accessing the fields directly from the `cloudCredsTemplateData` struct, so there is no reason to promote the members of the fields to the `cloudCredsTemplateData` struct.

Also, the fields should be pointers since they are optional.

```
type cloudCredsTemplateData struct {
   AWS *AWSCredsTemplateData
   OpenStack *OpenStackCredsTemplateData
   Libvirt *LibvirtCredsTemplateData
}
```

sallyom replied:
I was going to ask you about that yesterday, thanks, I thought I was missing something there

bison commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -39,8 +41,45 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
+			Base64encodeSecretAccessKey: base64.StdEncoding.EncodeToString([]byte(awscreds.SecretAccessKey)),
+		}
+	case installConfig.Config.Platform.Libvirt != nil:
+		libvirtCreds = Libvirt{
```
These will eventually be TLS credentials, i.e. a certificate and private key. The tricky part is that we probably can't fully manage the CA within the installer because you'll have to configure libvirtd with the CA as part of the initial manual setup.

So, at some point you'll have to either pass in a CA and key and generate the credentials here or just read a previously generated client certificate and key from disk and insert them.

PR #296 was the first pass at TLS for libvirt.

sallyom commented:
@staebler I squashed commits in order to fix merge conflicts (making it difficult to follow your review, I know).  I've addressed your feedback, though, ptal.  
- made secret data pointers, since they are optional
- renamed variables based on your suggestions
- separated the secret and role yamls, since the role yaml is same for all cloud-providers.  The secret data is what is unique, so each cloud-provider has a separate secret yaml.

sallyom commented:
/retest

staebler commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -74,6 +119,7 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
 	}
 
 	assetData := map[string][]byte{
+		"99_cloud-creds-secret.yaml":                             applyTemplateData(content.CloudCredsSecret, templateData),
```
We need a file for the cloud-creds-secret-reader role, too. Are you planning to add that back in with cloud-creds secret?

bparees commented:
/cc @legionus @coreydaley @dmage @gabemontero

sallyom commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -93,8 +93,6 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
 		}
-	default:
-		return errors.Errorf("unknown provider, could not populate cloud credentials secret data")
 	}
```
Don't throw error here, just generate an empty secret.  As we add different cloud-providers, they may/may not require credentials.  If they don't/we don't add any,  installer won't break if we don't default to error here.  It's not the responsibility to check for accepted cloud-provider here, anyways.

flaper87 commented:
@tomassedovic @hardys @russellb this change will allow us to upload a `clouds.yaml` into a secret on the cluster. A step forward towards enabling the cloud provider :+1: :tada:

dmage commented on pkg/asset/manifests/content/tectonic/cloud-creds-secret.go:
```diff
@@ -0,0 +1,24 @@
+  namespace: kube-system
+  name: cloud-creds
+data:
+{{- if .CloudCreds.AWS}}
```
Can we add the type of the cluster (AWS, OpenStack, etc.) so that consumers know what to expect?

bparees replied:
and while we're at it (providing AWS coordinates), the registry operator would also like to know what s3 bucket name should be used for the registry (if you don't supply a name we'll create one, but it's likely users may want to create their own bucket).

sallyom replied:
@dmage  working on it, yes

sallyom replied:
@bparees, makes sense, might not be within the scope of this PR, I'll discuss with the installer team.

sallyom commented on pkg/asset/manifests/content/tectonic/cloud-creds-secret.go:
```diff
@@ -0,0 +1,24 @@
+  aws_access_key_id: {{.CloudCreds.AWS.Base64encodeAccessKeyID}}
+  aws_secret_access_key: {{.CloudCreds.AWS.Base64encodeSecretAccessKey}}
+{{- else if .CloudCreds.OpenStack}}
+  credentials: {{.CloudCreds.OpenStack.Base64encodeCloudCreds}}
```
yup will do

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
/retest

sallyom commented:
@staebler @wking, I think this is ready for approval, ptal.  I've tested in libvirt, no secret/role is created and in aws `aws-creds` secret and `aws-creds-secret-reader` role are created

@flaper87 I've made the change to name os secret data `clouds.yaml` instead of `credentials`

flaper87 commented:
> @flaper87 I've made the change to name os secret data `clouds.yaml` instead of `credentials`

Perfect, thanks!

FWIW, I've tested this with openstack and https://github.com/openshift/release/pull/1824 It works fine!

wking commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -16,12 +20,17 @@ import (
 
 const (
 	tectonicManifestDir = "tectonic"
+	// TODO: Verify this is expected os creds file
```
[Looks like][1] we can drop this comment?

[1]: https://github.com/openshift/installer/pull/427/files#r227364281

wking commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -78,6 +116,18 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
+		}
+
+		for file, content := range conditionalAssetData {
+			assetData[file] = content
```
`conditionalAssetData` seems like more trouble than it's worth.  Why not just:

```go
assetData["99_..."] = applyTemplate...
```

sallyom replied:
because doh-duh thank you!

sallyom commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -16,12 +20,17 @@ import (
 	_ asset.WritableAsset = (*Tectonic)(nil)
+
+	// TODO: Verify which creds file to expect
+	//openStackCredsFile = os.Getenv("HOME") + "/.config/openstack"
```
:+1:  https://github.com/openshift/installer/issues/550

sallyom commented on pkg/asset/manifests/tectonic.go:
```diff
@@ -57,8 +66,37 @@ func (t *Tectonic) Generate(dependencies asset.Parents) error {
+	var cloudCreds cloudCredsSecretData
+	platform := ""
+	switch {
+	case installConfig.Config.Platform.AWS != nil:
```
:+1:  thanks, got it. I missed the addn of `func (p *Platform) Name() string` :)

sallyom commented:
/test gofmt

wking commented:
When you get a moment, can you squash the commits and drop the WIP?

```console
$ git log --oneline -2 origin/pr/427
f77fda4 trevor's feedback
47b84f0 WIP-add aws-creds-secret
```

is probably not a useful distinction for future readers ;).  While you're editing the commit message, pasting in whatever is relevant from [your initial PR comment][1] (e.g. "Various components will require access..." will make it easier for folks running `blame` to understand the motivation (otherwise they have to figure out the PR number and check here ;).

[1]: https://github.com/openshift/installer/pull/427#issue-220843466

sallyom commented:
@wking squashed, re-worded commit msg :)
